packages/core/src/Tool.ts

```diff
@@ -0,0 +1,107 @@
+/**
+ * A unique string identifier for each type of supported monorepo tool.
+ */
+export type ToolType = "bolt" | "lerna" | "pnpm" | "rush" | "yarn" | "none";
+
+/**
+ * A package.json access type.
+ */
+export type PackageAccessType = "public" | "restricted";
+
+/**
+ * An in-memory representation of a package.json file.
+ */
+export type PackageJSON = {
+    name: string;
+    version: string;
+    dependencies?: {
+        [key: string]: string;
+    };
+    peerDependencies?: {
+        [key: string]: string;
+    };
+    devDependencies?: {
+        [key: string]: string;
+    };
+    optionalDependencies?: {
+        [key: string]: string;
+    };
+    private?: boolean;
+    publishConfig?: {
+        access?: PackageAccessType;
+        directory?: string;
+        registry?: string;
+    };
+};
+
+/**
+ * A pre-loaded package json structure, with the diretory it was loaded from.
+ */
+export type Package = {
+    packageJson: PackageJSON;
+    dir: string;
+};
+
+/**
+ * A collection of packages, along with the monorepo tool used to load them,
+ * and (if supported by the tool) the associated "root" package.
+ */
+export type Packages = {
+    tool: Tool;
+    packages: Package[];
+    root?: Package;
+};
+
+/**
+ * An object representing the root of a specific monorepo, with the root
+ * directory and associated monorepo tool.
+ *
+ * Note that this type is currently not used by Tool definitions directly,
+ * but it is the suggested way to pass around a reference to a monorepo root
+ * directory and associated tool.
+ */
+export type MonorepoRoot = {
+    dir: string;
+    tool: Tool;
+};
+
+/**
+ * Monorepo tools may throw this error if a caller attempts to get the package
+ * collection from a directory that is not a valid monorepo root.
+ */
+export class InvalidMonorepoError extends Error { }
```

Comment:
We are constructing the message at all call sites but we could potentially construct it in the `constructor` here.

Reply:
If you'd really like to see this, I can push the template into the Error.

Right now I was relying more on convention because I expect the Tools to get more complex / detailed errors in some cases, e.g. `Directory blah is not a valid Blah monorepo: missing bolt property in package.json`, etc. So being able to pass in any string lets a Tool author customize the messaging, while still letting upstream code detect the entire class of error.

Reply:
Let's stick to a simple string